File: Media-Viewer-Module/Media-Viewer-Module/clsDataWork_Images.vb
Imports Ontology_Module
Imports OntologyClasses.BaseClasses

Public Class clsDataWork_Images
    Private objLocalConfig As clsLocalConfig
    Private objDBLevel_Images As clsDBLevel
    Private objDBLevel_Files As clsDBLevel
    Private objDBLevel_CreationDate As clsDBLevel
    Private dtblT_Images As New DataSet_Images.dtbl_ImagesDataTable
    Private objThread_Images As Threading.Thread
    Private objOItem_Ref As clsOntologyItem
    Private boolLoaded As Boolean
    Private boolTable As Boolean
    Private objLImages As List(Of clsMultiMediaItem) = New List(Of clsMultiMediaItem)

    Public ReadOnly Property Loaded As Boolean
        Get
            Return boolLoaded
        End Get
    End Property

    Public ReadOnly Property dtbl_Images As DataSet_Images.dtbl_ImagesDataTable
        Get
            Return dtblT_Images
        End Get
    End Property

    Public ReadOnly Property ItemList As List(Of clsMultiMediaItem)
        Get
            Return objLImages
        End Get
    End Property

    Public Sub get_Images(ByVal OItem_Ref As clsOntologyItem, Optional boolTable As Boolean = True)

        Me.boolTable = boolTable
        objOItem_Ref = OItem_Ref
        dtblT_Images.Clear()
        boolLoaded = False
        Try
            objThread_Images.Abort()
        Catch ex As Exception

        End Try


        objThread_Images = New Threading.Thread(AddressOf get_Images_Thread)
        objThread_Images.Start()

    End Sub

    Public Function GetNextOrderIDOFRef(OItem_Ref As clsOntologyItem) As Long
        Dim lngOrderID As Long
        Dim objOItem_Image As clsOntologyItem

        objOItem_Image = New clsOntologyItem
        objOItem_Image.GUID_Parent = objLocalConfig.OItem_Type_Images__Graphic_.GUID
        objOItem_Image.Type = objLocalConfig.Globals.Type_Object

        lngOrderID = objDBLevel_Images.get_Data_Rel_OrderID(objOItem_Image, OItem_Ref, objLocalConfig.OItem_RelationType_belongsTo, False)

        Return lngOrderID
    End Function

    Private Sub get_Images_Thread()
        Dim objOL_Images_To_Ref As New List(Of clsObjectRel)
        Dim objOL_Images_To_File As New List(Of clsObjectRel)
        Dim objOL_CreationDate As New List(Of clsObjectAtt)


        objOL_Images_To_Ref.Add(New clsObjectRel(Nothing, _
                                                 Nothing, _
                                                 objLocalConfig.OItem_Type_Images__Graphic_.GUID, _
                                                 Nothing, _
                                                 If(objOItem_Ref Is Nothing, Nothing, objOItem_Ref.GUID), _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing, _
                                                 objLocalConfig.OItem_RelationType_belongsTo.GUID, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing))

        objDBLevel_Images.get_Data_ObjectRel(objOL_Images_To_Ref, _
                                             boolIDs:=False)


        objOL_Images_To_File.Add(New clsObjectRel(Nothing, _
                                                  Nothing, _
                                                  objLocalConfig.OItem_Type_Images__Graphic_.GUID, _
                                                  Nothing, _
                                                  Nothing, _
                                                  Nothing, _
                                                  objLocalConfig.OItem_Type_File.GUID, _
                                                  Nothing, _
                                                  objLocalConfig.OItem_RelationType_belonging_Source.GUID, _
                                                  Nothing, _
                                                  objLocalConfig.Globals.Type_Object, _
                                                  Nothing, _
                                                  Nothing, _
                                                  Nothing))

        objDBLevel_Files.get_Data_ObjectRel(objOL_Images_To_File, _
                                            boolIDs:=False)

        objOL_CreationDate.Add(New clsObjectAtt(Nothing, _
                                                Nothing, _
                                                objLocalConfig.OItem_Type_Images__Graphic_.GUID, _
                                                objLocalConfig.OItem_Attribute_taking.GUID, _
                                                Nothing))

        objDBLevel_CreationDate.get_Data_ObjectAtt(objOL_CreationDate, _
                                                   boolIDs:=False)

        objLImages.Clear()
        objLImages = (From objImg In objDBLevel_Images.OList_ObjectRel
                         Join objFile In objDBLevel_Files.OList_ObjectRel On objFile.ID_Object Equals objImg.ID_Object
                         Group Join objDate In objDBLevel_CreationDate.OList_ObjectAtt On objDate.ID_Object Equals objImg.ID_Object Into objDates = Group
                         From objDate In objDates.DefaultIfEmpty
                         Order By objImg.OrderID
                         Select New clsMultiMediaItem(objImg.ID_Object, _
                                                      objImg.Name_Object, _
                                                      objImg.ID_Parent_Object, _
                                                      objFile.ID_Other, _
                                                      objFile.Name_Other, _
                                                      objFile.ID_Parent_Other, _
                                                      objDate, _
                                                      objImg.OrderID)).ToList()

        If boolTable Then
            For Each objImage In objLImages
                If objImage.OACreate Is Nothing Then
                    dtblT_Images.Rows.Add(objImage.OrderID, _
                                          objImage.ID_Item, _
                                          objImage.Name_Item, _
                                          Nothing, _
                                          objImage.ID_File, _
                                          objImage.Name_File)
                Else
                    dtblT_Images.Rows.Add(objImage.OrderID, _
                                          objImage.ID_Item, _
                                          objImage.Name_Item, _
                                          objImage.OACreate.Val_Date, _
                                          objImage.ID_File, _
                                          objImage.Name_File)
                End If
            Next
        End If


        boolLoaded = True
    End Sub

    Public Function GetImageItemOfFile(OItem_File As clsOntologyItem, OItem_MediaType As clsOntologyItem) As clsOntologyItem
        Dim objOItem_Result As clsOntologyItem
        Dim objOItem_MediaItem As clsOntologyItem
        Dim objOList_MediaItemToFile As New List(Of clsObjectRel)

        objOList_MediaItemToFile.Add(New clsObjectRel(Nothing, _
                                                      OItem_MediaType.GUID, _
                                                      OItem_File.GUID, _
                                                      Nothing, _
                                                      objLocalConfig.OItem_RelationType_belonging_Source.GUID,
                                                      objLocalConfig.Globals.Type_Object, _
                                                      Nothing, _
                                                      Nothing))


        objOItem_Result = objDBLevel_Images.get_Data_ObjectRel(objOList_MediaItemToFile, _
                                                                          boolIDs:=False)

        If objOItem_Result.GUID = objLocalConfig.Globals.LState_Success.GUID Then
            If objDBLevel_Images.OList_ObjectRel.Any() Then
                objOItem_MediaItem = New clsOntologyItem(objDBLevel_Images.OList_ObjectRel.First().ID_Object, _
                                                         objDBLevel_Images.OList_ObjectRel.First().ID_Other, _
                                                         objDBLevel_Images.OList_ObjectRel.First().ID_Parent_Other, _
                                                         objLocalConfig.Globals.Type_Object)


            Else
                objOItem_MediaItem = Nothing
            End If
        Else
            objOItem_MediaItem = Nothing
        End If

        Return objOItem_MediaItem
    End Function

    Public Function hasImage(OItem_Ref) As clsOntologyItem
        Dim objOL_Images_To_Ref As New List(Of clsObjectRel)

        objOL_Images_To_Ref.Add(New clsObjectRel(Nothing, _
                                                 Nothing, _
                                                 objLocalConfig.OItem_Type_Images__Graphic_.GUID, _
                                                 Nothing, _
                                                 OItem_Ref.GUID, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing, _
                                                 objLocalConfig.OItem_RelationType_belongsTo.GUID, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing, _
                                                 Nothing))

        Dim objOItem_Result = objDBLevel_Images.get_Data_ObjectRel(objOL_Images_To_Ref, doCount:=True)


        Return objOItem_Result
    End Function

    Public Function Rel_Image_To_File(OItem_Image As clsOntologyItem, OItem_File As clsOntologyItem) As clsObjectRel
        Dim objOR_Image_To_File = New clsObjectRel With {.ID_Object = OItem_Image.GUID, _
                                                             .ID_Parent_Object = OItem_Image.GUID_Parent, _
                                                             .ID_Other = OItem_File.GUID, _
                                                             .ID_Parent_Other = OItem_File.GUID_Parent, _
                                                             .ID_RelationType = objLocalConfig.OItem_RelationType_belonging_Source.GUID, _
                                                             .OrderID = 1, _
                                                             .Ontology = objLocalConfig.Globals.Type_Object}

        Return objOR_Image_To_File
    End Function

    Public Function Rel_Image_To_Ref(OItem_Image As clsOntologyItem, OItem_Ref As clsOntologyItem, Optional boolGetNextOrderID As Boolean = True) As clsObjectRel

        Dim intOrderID = OItem_Image.Level
        If boolGetNextOrderID Then
            intOrderID = objDBLevel_Images.get_Data_Rel_OrderID(New clsOntologyItem With {.GUID_Parent = OItem_Image.GUID_Parent}, OItem_Ref, objLocalConfig.OItem_RelationType_belongsTo, False)
            intOrderID = intOrderID + 1
        End If

        Dim objOR_Image_To_Ref = New clsObjectRel With {.ID_Object = OItem_Image.GUID, _
                                                            .ID_Parent_Object = OItem_Image.GUID_Parent, _
                                                            .ID_Other = OItem_Ref.GUID, _
                                                            .ID_Parent_Other = OItem_Ref.GUID_Parent, _
                                                            .ID_RelationType = objLocalConfig.OItem_RelationType_belongsTo.GUID, _
                                                            .OrderID = intOrderID, _
                                                            .Ontology = OItem_Ref.Type}

        Return objOR_Image_To_Ref
    End Function

    Public Function Rel_Image__Taking(OItem_Image As clsOntologyItem, dateTaking As DateTime) As clsObjectAtt
        Dim objOARel_Image__Taking = New clsObjectAtt With {.ID_AttributeType = objLocalConfig.OItem_Attribute_taking.GUID, _
                                                            .ID_Object = OItem_Image.GUID, _
                                                            .ID_Class = OItem_Image.GUID_Parent, _
                                                            .ID_DataType = objLocalConfig.Globals.DType_DateTime.GUID, _
                                                            .OrderID = 1, _
                                                            .Val_Named = dateTaking.ToString, _
                                                            .Val_Date = dateTaking}
        Return objOARel_Image__Taking
    End Function

    Public Sub New(ByVal LocalConfig As clsLocalConfig)
        objLocalConfig = LocalConfig
        set_DBConnection()
    End Sub

    Public Sub New(ByVal Globals As clsGlobals)
        objLocalConfig = New clsLocalConfig(Globals)
        set_DBConnection()
    End Sub

    Private Sub set_DBConnection()
        objDBLevel_Images = New clsDBLevel(objLocalConfig.Globals)
        objDBLevel_Files = New clsDBLevel(objLocalConfig.Globals)
        objDBLevel_CreationDate = New clsDBLevel(objLocalConfig.Globals)
    End Sub
End Class

File: ImageList-Module/ImageList-Module/clsDataWork_Images.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Ontology_Module;
using OntologyClasses.BaseClasses;
using Media_Viewer_Module;

namespace ImageList_Module
{

    public class clsDataWork_Images
    {
        private clsLocalConfig objLocalConfig;
        private clsOntologyItem objOItem_ImageList;

        private clsDBLevel objDBLevel_Images;
        private clsDBLevel objDBLevel_IsRoot;
        private clsDBLevel objDBLevel_KindOfOntologies;
        private clsDBLevel objDBLevel_Ref;

        public clsOntologyItem objOItem_Result_IsRoot { get; set; }
        public clsOntologyItem objOItem_Result_KindOfOntology { get; set; }
        public clsOntologyItem objOItem_Result_Images { get; set; }
        public clsOntologyItem objOItem_Result_ImageData { get; set; }
        public clsOntologyItem objOItem_Result_Ref { get; set; }

        private Media_Viewer_Module.clsDataWork_Images objDataWork_ImageData;
        private clsMediaItems objMediaItems;

        public List<clsImageOfImageList> ImageList { get; set; }

        public clsOntologyItem GetData_ImagesOfImageList(clsOntologyItem OItem_ImageList)
        {
            var objOItem_Result = objLocalConfig.Globals.LState_Success.Clone();

            objOItem_ImageList = OItem_ImageList;

            objOItem_Result_IsRoot = objLocalConfig.Globals.LState_Nothing.Clone();
            objOItem_Result_KindOfOntology = objLocalConfig.Globals.LState_Nothing.Clone();
            objOItem_Result_Images = objLocalConfig.Globals.LState_Nothing.Clone();
            objOItem_Result_ImageData = objLocalConfig.Globals.LState_Nothing.Clone();

            GetSubData_Images();
            objOItem_Result = objOItem_Result_Images;
            if (objOItem_Result.GUID == objLocalConfig.Globals.LState_Success.GUID)
            {
                GetSubData_IsRoot();
                objOItem_Result = objOItem_Result_IsRoot;

                if (objOItem_Result.GUID == objLocalConfig.Globals.LState_Success.GUID)
                {
                    GetSubData_KindOfOntologies();
                    objOItem_Result = objOItem_Result_KindOfOntology;

                    if (objOItem_Result.GUID == objLocalConfig.Globals.LState_Success.GUID)
                    {
                        GetSubData_Reference();
                        objOItem_Result = objOItem_Result_Ref;

                        if (objOItem_Result.GUID == objLocalConfig.Globals.LState_Success.GUID)
                        {
                            ImageList = (from objImage in objDBLevel_Images.OList_ObjectRel
                                         join objIsRoot in objDBLevel_IsRoot.OList_ObjectAtt on objImage.ID_Other equals objIsRoot.ID_Object into objIsRoots
                                         from objIsRoot in objIsRoots.DefaultIfEmpty()
                                         join objKindOfOntologies in objDBLevel_KindOfOntologies.OList_ObjectRel on objImage.ID_Other equals objKindOfOntologies.ID_Object into objKindsOfOntologies
                                         from objKindOfOntologies in objKindsOfOntologies.DefaultIfEmpty()
                                         join objRef in objDBLevel_Ref.OList_ObjectRel on objImage.ID_Other equals objRef.ID_Object into objRefs
                                         from objRef in objRefs.DefaultIfEmpty()
                                         select new clsImageOfImageList
                                         {
                                             ID_Image = objImage.ID_Other,
                                             Name_Image = objImage.Name_Other,
                                             ID_Attribute_IsRoot = objIsRoot != null ? objIsRoot.ID_Attribute : null,
                                             IsRoot = objIsRoot != null ? (bool)objIsRoot.Val_Bit : false,
                                             ID_KindOfOntology = objKindOfOntologies != null ? objKindOfOntologies.ID_Other : null,
                                             Name_KindOfOntology = objKindOfOntologies != null ? objKindOfOntologies.Name_Other : null,
                                             ID_Ref = objRef != null ? objRef.ID_Other : null,
                                             Name_Ref = objRef != null ? objRef.Name_Other : null,
                                             ID_Parent_Ref = objRef != null ? objRef.ID_Parent_Other : null,
                                             Type_Ref = objRef != null ? objRef.Ontology : null,
                                             ImageID = (long)objImage.OrderID
                                         }).ToList();

                            GetSubData_ImageData();

                            objOItem_Result_Images = objOItem_Result_ImageData;
                        }
                    }
                    

                   
                }
                
            }


            return objOItem_Result;
        }

        public void GetSubData_IsRoot()
        {
            objOItem_Result_IsRoot = objLocalConfig.Globals.LState_Nothing.Clone();

            if (objDBLevel_Images.OList_ObjectRel.Any())
            {
                List<clsObjectAtt> OARel_IsRoot;
                if (objDBLevel_Images.OList_ObjectRel.Count < 500)
                {
                    OARel_IsRoot = objDBLevel_Images.OList_ObjectRel.Select(i => new clsObjectAtt
                    {
                        ID_AttributeType = objLocalConfig.OItem_attributetype_is_root.GUID,
                        ID_Object = i.ID_Other
                    }).ToList();
                }
                else
                {
                    OARel_IsRoot = new List<clsObjectAtt> { new clsObjectAtt {ID_AttributeType = objLocalConfig.OItem_attributetype_is_root.GUID,
                        ID_Class = objLocalConfig.OItem_class_images__image_lists_.GUID}};
                }

                objOItem_Result_IsRoot = objDBLevel_IsRoot.get_Data_ObjectAtt(OARel_IsRoot, boolIDs: false);

            }
            else
            {
                objOItem_Result_IsRoot = objLocalConfig.Globals.LState_Success.Clone();
            }
        }

        public void GetSubData_KindOfOntologies()
        {
            objOItem_Result_KindOfOntology = objLocalConfig.Globals.LState_Nothing.Clone();

            if (objDBLevel_Images.OList_ObjectRel.Any())
            {
                List<clsObjectRel> ORel_KindOfOntology;
                if (objDBLevel_Images.OList_ObjectRel.Count < 500)
                {
                    ORel_KindOfOntology = objDBLevel_Images.OList_ObjectRel.Select(i => new clsObjectRel
                    {
                        ID_Object = i.ID_Other,
                        ID_RelationType = objLocalConfig.OItem_relationtype_reference_to.GUID,
                        ID_Parent_Other = objLocalConfig.OItem_class_kindofontology.GUID
                    }).ToList();
                }
                else
                {
                    ORel_KindOfOntology = new List<clsObjectRel> { new clsObjectRel 
                    {
                        ID_Parent_Object = objLocalConfig.OItem_class_images__image_lists_.GUID,
                        ID_RelationType = objLocalConfig.OItem_relationtype_reference_to.GUID,
                        ID_Parent_Other = objLocalConfig.OItem_class_kindofontology.GUID
                    }};
                }

                objOItem_Result_KindOfOntology = objDBLevel_KindOfOntologies.get_Data_ObjectRel(ORel_KindOfOntology, boolIDs: false);

            }
            else
            {
                objOItem_Result_KindOfOntology = objLocalConfig.Globals.LState_Success.Clone();
            }
        }

        public void GetSubData_Reference()
        {
            objOItem_Result_Ref = objLocalConfig.Globals.LState_Nothing.Clone();

            if (objDBLevel_Images.OList_ObjectRel.Any())
            {
                List<clsObjectRel> ORel_Ref;
                if (objDBLevel_Images.OList_ObjectRel.Count < 500)
                {
                    ORel_Ref = objDBLevel_Images.OList_ObjectRel.Select(i => new clsObjectRel
                    {
                        ID_Object = i.ID_Other,
                        ID_RelationType = objLocalConfig.OItem_relationtype_belonging_resource.GUID
                    }).ToList();
                }
                else
                {
                    ORel_Ref = new List<clsObjectRel> { new clsObjectRel 
                    {
                        ID_Parent_Object = objLocalConfig.OItem_class_images__image_lists_.GUID,
                        ID_RelationType = objLocalConfig.OItem_relationtype_belonging_resource.GUID
                    }};
                }

                objOItem_Result_Ref = objDBLevel_Ref.get_Data_ObjectRel(ORel_Ref, boolIDs: false);

            }
            else
            {
                objOItem_Result_Ref = objLocalConfig.Globals.LState_Success.Clone();
            }
        }

        public void GetSubData_Images()
        {
            objOItem_Result_Images = objLocalConfig.Globals.LState_Nothing.Clone();

            var objORel_Images = new List<clsObjectRel> { new clsObjectRel {ID_Object = objOItem_ImageList.GUID,
                ID_RelationType = objLocalConfig.OItem_relationtype_contains.GUID,
                ID_Parent_Other = objLocalConfig.OItem_class_images__image_lists_.GUID } };

            objOItem_Result_Images = objDBLevel_Images.get_Data_ObjectRel(objORel_Images, boolIDs: false);

        }


        public void GetSubData_ImageData()
        {
            objOItem_Result_ImageData = objLocalConfig.Globals.LState_Nothing.Clone();

            foreach (var image in ImageList)
            {
                objDataWork_ImageData.get_Images(new clsOntologyItem
                {
                    GUID = image.ID_Image,
                    Name = image.Name_Image,
                    GUID_Parent = objLocalConfig.OItem_class_images__image_lists_.GUID,
                    Type = objLocalConfig.Globals.Type_Object
                }, false);

                while (!objDataWork_ImageData.Loaded) ;

                var objImageDatas = objDataWork_ImageData.ItemList;

                var objOItem_Result = GetSubData_ImageFile(image, objImageDatas);
                if (objOItem_Result.GUID == objLocalConfig.Globals.LState_Error.GUID)
                {
                    objOItem_Result_ImageData = objOItem_Result;
                    break;
                }
            }

            if (objOItem_Result_ImageData.GUID == objLocalConfig.Globals.LState_Nothing.GUID)
            {
                objOItem_Result_ImageData = objLocalConfig.Globals.LState_Success.Clone();
            }
        }

        public clsOntologyItem GetSubData_ImageFile(clsImageOfImageList objImage, List<clsMultiMediaItem> mediaItems)
        {
            var objOItem_Result = objLocalConfig.Globals.LState_Success.Clone();

            if (mediaItems.Any())
            {
                objImage.ImageData = objMediaItems.Get_ImageFile(new clsOntologyItem
                {
                    GUID = mediaItems.First().ID_File,
                    Name = mediaItems.First().Name_File,
                    GUID_Parent = mediaItems.First().ID_Parent_File,
                    Type = objLocalConfig.Globals.Type_Object
                });


            }

            return objOItem_Result;
        }

        public clsDataWork_Images(clsLocalConfig LocalConfig)
        {
            objLocalConfig = LocalConfig;

            Initialize();
        }

        private void Initialize()
        {
            objDBLevel_Images = new clsDBLevel(objLocalConfig.Globals);
            objDBLevel_IsRoot = new clsDBLevel(objLocalConfig.Globals);
            objDBLevel_KindOfOntologies = new clsDBLevel(objLocalConfig.Globals);
            objDBLevel_Ref = new clsDBLevel(objLocalConfig.Globals);

            objDataWork_ImageData = new Media_Viewer_Module.clsDataWork_Images(objLocalConfig.Globals);
            objMediaItems = new clsMediaItems(objLocalConfig.Globals);
        }

    }
}
